refactor (shell) : Refactor CRC unix/linux shell detection logic using gopsutil (#4562)

rohanKanojia opened the pull request:
## Description

<!--
Thank you for your pull request (PR)!

Please provide a description of what your PR does providing a link (if applicable) to the issue it fixes.
-->

Fixes: #4562 

Relates to: https://github.com/crc-org/crc/pull/4526#issuecomment-2579431999

Refactor CRC Shell detection for UNIX/Linux Systems to rely on parent process name. I was only able to get this approach working for Linux/Unix systems where gopsutil library is correctly detecting parent PIDs. On WSL, it's not able to see beyond windows processes.

<!--
Describe in plain English what you solved and how. For instance, _Added `start` command to CRC so the user can create a VM and set up a single-node OpenShift cluster on it with one command. It requires blablabla..._
-->

## Type of change
<!--
What types of changes does your code introduce? Put an `x` in all the boxes that apply
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change
- [x] Chore (non-breaking change which doesn't affect codebase;
  test, version modification, documentation, etc.)

## Proposed changes
<!--
List main as well as consequential changes you introduced or had to introduce.

1. Add `start` command.
2. Add `setup` as prerequisite to `start`.
3. ...
-->
+ Move parsing ps output and picking recent pid logic to shell_windows. It would only be used in case of WSL.
+ Instead of parsing ps output and picking up recent pid. Inspect parent process of current process and keep going up until we find a shell process. Pick that shell process as currently active shell.
    + As of now I've set a depth limit of 10 levels, if we're not able to find any shell process after going this deep, we return blank value.



## Testing
<!--
What is the _bottom-line_ functionality that needs testing? Describe in pseudo-code or in English. Use verifiable statements that tie your changes to existing functionality.

1. `start` succeeds first time after `setup` succeeded
2. stdout contains ... if `start` succeeded
3. stderr contains ... after `start` failed
4. `status` returns ... if `start` succeeded
5. `status` returns ... if `start` failed
6. `start` fails after `start` succeeded or after `status` says CRC is `Running`
7. ...
-->
Start CRC cluster and run `oc-env` / `podman-env` commands on bash, zsh or fish shells.

## Contribution Checklist
- [x] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
- [x] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Which platform have you tested the code changes on? <!-- Only put an `x` in applicable platforms -->
    - [x] Linux
    - [x] Windows
    - [x] MacOS (QE Lab instance)

rohanKanojia committed:
refactor (shell) : Refactor unix/linux shell detection mechanism (#4562)

+ Move parsing ps output and picking recent pid approach to shell_windows. It would only
  be used in case of WSL.
+ Instead of parsing ps output and picking up recent pid. Inspect parent process
  of current process and keep going up until we find a shell process. Pick that
  shell process as currently active shell.

Signed-off-by: Rohan Kumar <rohaan@redhat.com>

cfergeau commented:
> On WSL, it's not able to see beyond windows processes.

What about Windows with powershell or cmd?

rohanKanojia commented:
@cfergeau : In this PR I had kept changes limited to parts I had touched on in https://github.com/crc-org/crc/pull/4526 . In my previous pull request for Shell detection, I had only made changes for WSL and Unix, Linux platforms.

I just tested this approach on PowerShell and CMD. I can verify that this approach is able to detect shell processes from parent tree.

Some Debug logs on PowerShell:
```
PS C:\Users\rokum\go\src\github.com\crc-org\crc> C:\Users\rokum\go\bin\crc.exe oc-env
process name : crc.exe
parent process id : 8060
process name : powershell.exe
parent process id : 15508
process name : WindowsTerminal.exe
parent process id : 8592
process name : explorer.exe
parent process id : 8476
process name :
 detectShellByCheckingProcessTree
$Env:PATH = "C:\Users\rokum\.crc\bin\oc;$Env:PATH"
# Run this command to configure your shell:
# & crc oc-env | Invoke-Expression
```

Some Debug logs on CMD:
```
C:\Users\rokum>C:\Users\rokum\go\bin\crc.exe oc-env
process name : crc.exe
parent process id : 14804
process name : cmd.exe
parent process id : 8592
process name : explorer.exe
parent process id : 8476
process name :
 detectShellByCheckingProcessTree
SET PATH=C:\Users\rokum\.crc\bin\oc;%PATH%
REM Run this command to configure your shell:
REM @FOR /f "tokens=*" %i IN ('crc oc-env') DO @call %i
```

Shall I modify this PR to cover PowerShell and CMD as well?

cfergeau commented on pkg/os/shell/shell_windows.go:
```diff
@@ -101,3 +105,83 @@ func detect() (string, error) {
 }
+
+// detectShellByInvokingCommand is a utility method that tries to detect current shell in use by invoking `ps` command.
+// This method is extracted so that it could be used by unix systems as well as Windows (in case of WSL). It executes
```
I think now it is only used on Windows in the WSL case?

rohanKanojia replied:
Sorry, I forgot to update this comment after moving the method. Thanks for noticing! It should be okay now.

cfergeau commented on pkg/os/shell/shell_unix.go:
```diff
@@ -6,20 +6,99 @@ package shell
 var (
-	ErrUnknownShell = errors.New("Error: Unknown shell")
+	ErrUnknownShell                             = errors.New("Error: Unknown shell")
+	getCurrentProcessID                         = os.Getpid
```
In my opinion, it is slightly easier to follow if you name this `getPid`, in a way this removes one level of indirection.

rohanKanojia replied:
Thanks, I've renamed the getter to `getPid` as you suggested.

cfergeau commented on pkg/os/shell/shell_unix.go:
```diff
@@ -6,20 +6,99 @@ package shell
+	"os"
 	"path/filepath"
+
+	"github.com/shirou/gopsutil/v3/process"
```
Why not v4? We probably should switch crc to the latest version, but this can be done in a follow-up commit if you prefer.

rohanKanojia replied:
I've added a new commit on top of this one to upgrade to v4

cfergeau commented on pkg/os/shell/shell_unix.go:
```diff
@@ -6,20 +6,99 @@ package shell
+type AbstractProcess interface {
+	Name() (string, error)
+	Ppid() (int32, error)
+}
```
It seems this code could be simpler if you use [`Parent()`](https://pkg.go.dev/github.com/shirou/gopsutil/v4@v4.24.12/process#Process.Parent) instead of `Ppid()` ?

And I think it's possible to remove `	getCurrentProcessID                         = os.Getpid` and `AbstractProcessSupplier` and to replace both with `var newProcess = process.NewProcess`.

Or will this prevent some of the "failure" tests that you have from being doable?

cfergeau replied:
Regardless of the previous suggestion, I'd add a comment saying these types are needed for the unit tests.
You could add `var _ AbstractProcess = process.Process` to make it clear that this type implements the interface.

rohanKanojia replied:
Thanks for the suggestion, I've updated it to use [`Parent()`](https://pkg.go.dev/github.com/shirou/gopsutil/v4@v4.24.12/process#Process.Parent), as you suggested. It's much clearer now!

cfergeau replied:
Ah, the fact that `Parent` must return an `AbstractProcess` makes things a bit more complicated than I hoped as a wrapper is needed for `*process.Process`, but I agree it is simpler anyway.

cfergeau reviewed (commented):
Another potential follow-up is to use `detectShellByCheckingProcessTree` in `detect()` in `shell_windows.go` as both seem to have similar functionalities.

rohanKanojia commented:
> Another potential follow-up is to use detectShellByCheckingProcessTree in detect() in shell_windows.go as both seem to have similar functionalities.

Can I tackle this one as a follow-up task?

rohanKanojia committed:
chore (deps) : Upgrade gopsutil library from v3.24.5 to v4.24.12

We're currently using an old version of gopsutil library. Upgrade to the latest
version in order keep dependency up to date. This commit includes changes to update the
dependency version in project and also vendor directory updates

Signed-off-by: Rohan Kumar <rohaan@redhat.com>

cfergeau commented:
> > Another potential follow-up is to use detectShellByCheckingProcessTree in detect() in shell_windows.go as both seem to have similar functionalities.
> 
> Can I tackle this one as a follow-up task?

Yup definitely!

cfergeau commented on pkg/os/shell/shell_unix.go:
```diff
@@ -6,20 +6,100 @@ package shell
 
 var (
 	ErrUnknownShell = errors.New("Error: Unknown shell")
+	getPid          = getCurrentProcess
```
With the changes you did, `getPid` is no longer very accurate :-/

rohanKanojia replied:
Sorry for the leftover typo, I've renamed it.

cfergeau commented on pkg/os/shell/shell_unix.go:
```diff
@@ -6,20 +6,100 @@ package shell
+// process.Process object. This implementation is used in production code.
+type RealProcess struct {
+	realProcess *process.Process
+}
```
If you don't give a name to `realProcess`, you don't need to reimplement `Name`, go will reuse the implementation from `*process.Process`:
```
type RealProcess struct {
	*process.Process
}
```

rohanKanojia replied:
Thanks, I've updated it as per your suggestion.

cfergeau commented on pkg/os/shell/shell_unix.go:
```diff
@@ -6,20 +6,100 @@ package shell
+	return &RealProcess{parentProcess}, nil
+}
+
+func NewRealProcess(pid int32) (*RealProcess, error) {
```
`getCurrentProcess` is the single user of `NewRealProcess`, you could consider merging the 2 especially as most of their code is error-handling.

rohanKanojia replied:
You're right. I've merged both methods into single one

cfergeau reviewed (commented):
minor comments left, looks good to me!

cfergeau commented:
/lgtm
/approve
